app/src/main/java/com/android/sample/model/workout/WorkoutRepositoryFirestore.kt

```diff
@@ -229,6 +230,175 @@ open class WorkoutRepositoryFirestore<T : Workout>(
+                                as Map<String, Any>?)!!
+                        // Delete the workout id from the done list of the user
+                        db.collection(collectionPath)
+                            .document(doneDocumentName)
+                            .collection(documentName)
+                            .document(id)
+                            .delete()
+                            .addOnSuccessListener { onSuccess() }
+                            .addOnFailureListener { onFailure(it) }
+                        // Add the id to the user workout list
+                        db.collection(collectionPath)
+                            .document(documentToCollectionName)
+                            .collection(documentName)
+                            .document(id)
+                            .set(dataMapWorkoutID)
+                            .addOnFailureListener { e -> onFailure(e) }
+                      }
+                      .addOnFailureListener { onFailure(it) }
+                }
+                .addOnFailureListener { onFailure(it) }
+          } else {
+            onFailure(Exception("Document with ID $id does not exist in done"))
+          }
+        }
+        .addOnFailureListener { onFailure(it) }
+  }
+
+  /**
+   * Retrieves done documents from the Firestore.
+   *
+   * @param onSuccess A callback function that is invoked with the workout data upon successful
+   *   retrieval.
+   * @param onFailure A callback function that is invoked in case of an error during the retrieval.
+   */
+  override fun getDoneDocuments(onSuccess: (List<T>) -> Unit, onFailure: (Exception) -> Unit) {
+    // we first get document'ids from user done documents
+    db.collection(collectionPath)
+        .document(doneDocumentName)
+        .collection(documentName)
+        .get()
+        .addOnCompleteListener { task ->
+          if (task.isSuccessful) {
+            val workoutids =
+                task.result?.mapNotNull { document ->
+                  documentSnapshotToObject(document, adapterWorkoutID) { it?.workoutid } as? String
+                } ?: emptyList()
+            val workouts = mutableListOf<T>()
+            val tasks = mutableListOf<Task<DocumentSnapshot>>()
+
+            // for each id we get the content of the workout in "doneWorkouts" document
+
+            for (id in workoutids) {
+              val task =
+                  db.collection(doneDocumentName)
+                      .document(id)
+                      .get()
+                      .addOnSuccessListener { document ->
+                        val workout = documentSnapshotToObject(document, adapter) as T
+                        workout.let { workouts.add(workout) }
+                      }
+                      .addOnFailureListener { e ->
+                        Log.e(
+                            "WorkoutRepositoryFirestore", "Error getting done workout document", e)
```

Comment:
Consider extracting the logTag into a separate constant.

Reply:
You're right, done.